### PR TITLE
Api & connection to the front end

### PR DESCRIPTION
**Pull request by PaulConyngham**
many various fix's changes and mods to get the spagbol package, exposed and api & connected to Ardhi's pre-existing front end

### CODE REVIEW

**Comment by ardhihdra** [comment]
for the backend I can't run it through main.py, which will be the main entry file moving forward? 
other than that I am reviewing and learning the backend atm as I haven't seen it in a while

**Comment by PaulConyngham** [comment]
Thanks Ardhi, you can run the api from root using: 

python -m spagbol.api.spagbol_api

**Review by ardhihdra — APPROVED** [review]

File: spagbol/api/modules.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from injector import Module, provider, singleton
```
Comment (AndriiVelykorodnii) [inline spagbol/api/modules.py:1]: It's better to separate local modules with installed modules during the import.

File: spagbol/controllers/spagbol_controller.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+"""
+
+import logging
+from spagbol.loading import DataLoader, AlpacaLoader
+from spagbol.embedding import Embedder, AllMiniLMEmbedder
+from spagbol.clustering import ClusteringModel, GaussianMixtureClustering
+from spagbol.reduction import DimensionalityReduction, PcaReduction
+from spagbol.errors import InvalidSourceError, NoDatasetError, ClusteringError
+
+class SpagbolController:
```
Comment (AndriiVelykorodnii) [inline spagbol/controllers/spagbol_controller.py:18]: The name is misleading. It's not a controller, better to call it SpagbolPipeline, IMHO.

File: spagbol/embedding/AllMiniLMEmbedder.py [hunk 3]
```diff
@@ -78,26 +82,35 @@ def embed(self, data: str) -> np.array:
...
-
-            embeddings = F.normalize(embeddings, p=2, dim=1)
-
-            return embeddings.detach().numpy()
-        except Exception as e:
-            print(f"Error embedding batch data: {e}")
-            return None
+        embeddings_list = []
+        # Process each sentence individually and update the progress bar
+        for sentence in tqdm(data, desc="Embedding sentences"):
```
Comment (AndriiVelykorodnii) [inline spagbol/embedding/AllMiniLMEmbedder.py:96]: Why would you change the batched input to just looped input? Batched would be significantly faster. If memory is our concern we could just run smaller batches.

File: spagbol/spagbol.py [hunk 4]
```diff
@@ -32,18 +36,118 @@ def __init__(self, data_loader: DataLoader, embedder: Embedder, clustering_model
...
+        json_data = json.dumps(data_for_json, indent=4)
+        
+        return json_data
 
     def find_similarities(self):
         pass
 
+    def get_data_points(self, criteria: Dict[str, Any]) -> List[Dict[str, Any]]:
+        # Filter the dataset based on the criteria
+        filtered_data = self.dataset[self.dataset.columns.intersection(criteria.keys())].isin(criteria).dropna()
```
Comment (AndriiVelykorodnii) [inline spagbol/spagbol.py:134]: I think it would be better to separate some of these operations and not stack them in one line

**Review by AndriiVelykorodnii — APPROVED** [review]
One additional overall comment. I think, it's better to reduce number of line comments in code, it's not necessary to comment every line, in my opinion. Otherwise, looks good.

**Comment by PaulConyngham** [comment]
@AndriiVelykorodnii Will make your suggested changes in a future pull request

**Pull request merged**